app/src/main/java/de/westnordost/streetcomplete/data/elementfilter/BooleanExpression.kt

```diff
@@ -94,6 +97,28 @@ class AnyOf<I : Matcher<T>, T> : Chain<I, T>() {
     override fun toString() = nodes.joinToString(" or ") { "$it" }
 }
 
+class Not<I : Matcher<T>, T> : Chain<I, T>() {
+    override fun addChild(child: BooleanExpression<I, T>) {
+        if (nodes.size > 0)
+        {
+            throw IllegalStateException("Adding a second child to '!' (NOT) operator is not allowed")
+        }
+        super.addChild(child)
+    }
+
+    override fun matches(obj: T) = !nodes.first().matches(obj)
+
+    override fun toString() : String {
+        val childString = when (val firstNode = nodes.firstOrNull()) {
+            is Chain -> "($firstNode)"
+            null -> ""
+            else -> "($firstNode)"
+        }
+
+        return "!$childString"
+    }
+}
+
```

Comment:
Looks like this can be simplified to

```suggestion
    override fun toString() = "!(${nodes.firstOrNull() ?: ""})"
```